### ISSUE TITLE
V2 switch to cumulative-only display is not obvious + rate-limiting for currency API is too tight (calc takes quite a while)

### ISSUE THREAD

**Issue by vitich** [mxmCherry/taxes#7]
v2 incorrectly calculates my income...
And the execution time instead of 2-3 sec became 20-30 sec 
I have a single config file since the summer of 2018
:(

**Comment by mxmCherry** [comment]
> And the execution time instead of 2-3 sec became 20-30 sec
> I have a single config file since the summer of 2018

https://github.com/mxmCherry/taxes#notes I added rate-limiter to pull them, DoS-ing free APIs is no good. Just comment out prev years' data and it'll become faster.

> v2 incorrectly calculates my income...

This is a bad, very bad bug report, it tells me nothing. Make some synthetic test cases to show me what's wrong (it worked fine for my data for a couple of years).

UPD: I should admit, it wasn't 100% fine because of precision loss, up to 1-2 kopeck diff with v2 for multiple transactions per quarter (because now every foreign-currency transaction is rounded to 2 decimals if you set rounding precision - I do set it for my data).

**Comment by mxmCherry** [comment]
You can even write/PR your own test (the `It(...)` thing) here: https://github.com/mxmCherry/taxes/blob/master/internal/tax/tax_test.go

**Comment by vitich** [comment]
I am sorry for my bad bug report. 
Look at the Q4 difference on the screenshot.
Please pay attention in the 4th quarter the income is in both dollars and euros. In the 3rd quarter there were only dollars and counted normally.
Also, please pay attention to the execution time under the same conditions:
v1 - 0.292 sec
v2 - 3.041 sec
I hope this bug report is better )))
![taxes_diff](https://user-images.githubusercontent.com/12166319/149230906-1f6e5f58-0aa4-4a42-88e9-75d285fee154.png)

**Comment by vitich** [comment]
Do you know why 62792.45 in v2? )
13464.90 + 49327.55 = 62792.45
It's funny but why does it add Q3 and Q4 (v1) and show me as Q4 (v2)?

**Comment by mxmCherry** [comment]
@vitich that's the intended behavior: https://github.com/mxmCherry/taxes#using

> This tool prints (STDOUT) simplified table output with cumulative income/taxes - since beginning of the year, so Q2 already includes Q1, Q3 includes Q2 and so on.

I was tired to bother adding them for each report, as that's how it is declared - a full amount since beginning of the year.

Makes sense, no "bug" then?

Regarding timings - commented above. That's how it'll work, it'll always take a few seconds because of rate limiting requests to currency rates API.

**Comment by mxmCherry** [comment]
Just in case - right now I'm trying to confirm there's no real bug with calcs, only misunderstanding (as now it outputs cumulative sums from beginning of the year).

I think, the following can be done (maybe over w/e):

- print out a warning/note that sums are cumulative (calced from beginning of the year) so it's less misleading
- maybe lower rate-limiting - allow one req per 200-300ms (so 5 RPS or so)
- maybe make rate-limiter configurable as it's hardcoded right now, not sure where to do that though (better in YAML, but that's a bit more complex than just adding a flag to CLI)

**Comment by vitich** [comment]
Well ... the functionality of the previous version suited me very much because I used your program as a calculator for quarterly reports for the tax office and as an e-book of income. 
If you display separately the 3rd and 4th quarter, then I would like to see the individual values. And the final cumulative sum of all quarters could be separately.
But, of course, you are the author of the program and you know better what you need.
In any case, thank you very much for these several years of using the first version of the program.

**Comment by mxmCherry** [comment]
I think, I'll better add additional columns to output, so it becomes like:

```
... | Quarter Income | Quarter Tax | Year Income (Cumulative) | Year Tax (Cumulative)
```

I just thought that since we declare only cumulative amount, it's fine to expose only cumulative data.

But I think it won't hurt to have more details in output.

That's a task for upcoming (or next) weekend, along with rate-limiting tweaks (200ms delay should be fine, it'll be like 5 times faster already).

I'll keep this issue open to track these 2 things.

**Comment by mxmCherry** [comment]
Changed request-wait-interval 500ms -> 150ms, should become a bit faster as of `v2.0.0-alpha.5`:

```shell
go install github.com/mxmCherry/taxes/v2@v2.0.0-alpha.5
```

Quarterly (non-cumulative income) to be addressed a bit later, by adding `Q Income | Q Tax | Y Income | Y Tax` columns (Q - per-quarter, Y - cumulative / since beginning of the year).

This bit requires a few additions in main calc logic/data-objects (right now Quarter stores only cumulative income/tax, need to break out the per-quarter amounts).

**Comment by mxmCherry** [comment]
Released individual quarter amount exposing as `v2.0.0-beta.1`:

```shell
go install github.com/mxmCherry/taxes/v2@v2.0.0-beta.1
```

Calling it beta now - don't think much (if anything) will change in current implementation. Planning to tag a stable v2.0.0 once I use this tool for Q1 tax declaration in April with manual re-calcs (to be 200% sure it's all fine).
